Composer/packages/server/src/models/connector/csharpBotConnector.ts

```diff
@@ -2,74 +2,168 @@
-import Path from 'path';
+import { ChildProcess, spawn } from 'child_process';
 
-import axios from 'axios';
 import archiver from 'archiver';
-import FormData from 'form-data';
 
 import { BotProjectService } from '../../services/project';
 import { DialogSetting } from '../bot/interface';
+import { Path } from '../../utility/path';
+import log from '../../logger';
 
 import { BotConfig, BotEnvironments, BotStatus, IBotConnector, IPublishHistory } from './interface';
 
 export class CSharpBotConnector implements IBotConnector {
-  private adminEndpoint: string;
+  public status: BotStatus = BotStatus.NotConnected;
   private endpoint: string;
-  constructor(adminEndpoint: string, endpoint: string) {
-    this.adminEndpoint = adminEndpoint;
+  private runtime: ChildProcess | null = null;
+  constructor(endpoint: string) {
     this.endpoint = endpoint;
+    this.addProcessListeners();
   }
 
-  public status: BotStatus = BotStatus.NotConnected;
+  private addProcessListeners = () => {
+    process.on('SIGINT', () => {
+      log('[SIGINT] start graceful shutdown');
+      this.stop();
+      process.exit(1);
+    });
+    process.on('SIGTERM', () => {
+      log('[SIGTERM] start graceful shutdown');
+      this.stop();
+      process.exit(1);
+    });
+    process.on('SIGQUIT', () => {
+      log('[SIGQUIT] start graceful shutdown');
+      this.stop();
+      process.exit(1);
+    });
+  };
 
-  connect = async (_: BotEnvironments, __: string) => {
-    // confirm bot runtime is listening here
-    try {
-      await axios.get(this.adminEndpoint + '/api/admin');
-    } catch (err) {
-      throw new Error(err);
+  private stop = () => {
+    if (this.runtime) {
+      log(`kill this bot with process PID: ${this.runtime.pid}`);
+      this.runtime.kill('SIGKILL');
+      this.runtime = null;
     }
-
     this.status = BotStatus.NotConnected;
+  };
 
-    return `${this.endpoint}/api/messages`;
+  private buildProcess = async (dir: string): Promise<number | null> => {
+    return new Promise((resolve, reject) => {
+      const startScript = Path.resolve(dir, './Scripts/build_runtime.ps1');
+      const build = spawn(`pwsh ${startScript}`, {
+        cwd: dir,
+        detached: true,
+        shell: true,
+        stdio: ['ignore', 'ignore', 'inherit'],
+      });
+      log(`build pid : ${build.pid}`);
+
+      build.stderr &&
+        build.stderr.on('data', function(err) {
+          reject(err.toString());
+        });
+
+      build.on('exit', function(code) {
+        resolve(code);
+      });
+    });
   };
 
-  sync = async (config: DialogSetting) => {
-    // archive the project
-    // send to bot runtime service
-    const currentProject = BotProjectService.getCurrentBotProject();
-    if (currentProject === undefined) {
-      throw new Error('no project is opened, nothing to sync');
+  private getConnectorConfig = (config: DialogSetting) => {
+    const configList: string[] = [];
+    if (config.MicrosoftAppPassword) {
```

Comment:
can we make the three sensitive config configable? and we should also include qna:endpointKey here.

Reply:
done, If user set SensitiveProperties in settings, if will be filter too.